### PR TITLE
Add a new vagrant provider: OpenStack

### PR DESCRIPTION
**Pull request by martbhell**
Hello!

Only tested with OpenStack Newton with Neutron and floating IPs.
After this change only tested with a VirtualBox install and that launched 
a virtualbox VM but I haven't spent time setting up the networking so it 
didn't get very far.

I didn't write anything into the root/README.md about this provider as it's with this PR only available with the icinga2x repo. Would be nice if the other providers could be tested with this Vagrantfile to make sure it still works.

Needed to move the vm.box setting into each provider as the openstack
provider didn't like it.

Make this work on CentOS.

Write director's kickstart.ini with a template (to get actual hostname
in there)

For openstack provider print the @@ssh_ip@@ at the end (too).

Usage in README.md but basically set some environment variables and
press play :)

### CODE REVIEW

**Comment by dnsmichi** [comment]
I'd be glad if anyone else who has OpenStack available could test this PR.

**Comment by martbhell** [comment]
If someone with the other providers available could test some/those too that would be great :)

File: icinga2x/Vagrantfile [hunk 1]
```diff
@@ -36,7 +37,6 @@ nodes = { 'icinga2' => {
 Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
   nodes.each_pair do |name, options|
     config.vm.define name do |node_config|
-      node_config.vm.box = options[:box_virtualbox]
```
Comment (dnsmichi) [inline icinga2x/Vagrantfile]: Removing this default will break other providers, if there is no openstack plugin installed.

File: icinga2x/Vagrantfile [hunk 2]
```diff
@@ -65,6 +85,8 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
 
       # provider: virtualbox
       node_config.vm.provider :virtualbox do |vb, override|
+        node_config.ssh.private_key_path = "~/.ssh/id_rsa"
```
Comment (dnsmichi) [inline icinga2x/Vagrantfile]: Does this change make sense being added to other providers too? It doesn't seem related to OpenStack only.

File: modules/icingaweb2/templates/kickstart.ini.erb [hunk 3]
```diff
@@ -1,5 +1,5 @@
 [config]
-endpoint = icinga2
+endpoint = <%= @fqdn %>
```
Comment (dnsmichi) [inline modules/icingaweb2/templates/kickstart.ini.erb]: Should be a separate commit next time, at first glance it wasn't clear to me why it is needed here.

File: scripts/shell_provisioner.sh [hunk 4]
```diff
@@ -53,7 +53,7 @@ elif [ "$OSTYPE" = "Ubuntu" ]; then
         rm -f /tmp/puppetlabs.deb
         apt-get update
     fi
-elif [ "$OSTYPE" = "RedHat" ]; then
+elif [ "$OSTYPE" = "RedHat" ]  || [ "$OSTYPE" = "CentOS" ]; then
```
Comment (dnsmichi) [inline scripts/shell_provisioner.sh]: CentOS sets the OSTYPE to RedHat too, where is that change coming from? I just want to be sure about it.

**Comment by dnsmichi** [comment]
Did a short review to fully understand its purpose. I don't like the ENV variables, since they change the behaviour how you configure Vagrant, but it is a start for anything else. I was considering some config file, but feared to add yet another Ruby/Vagrant dependency just for users who do `vagrant up` on Windows.

File: scripts/shell_provisioner.sh (hunk 4 above)
Comment (martbhell) [inline scripts/shell_provisioner.sh]: In all my centos servers (physical, virtual, 6, 7) it looks like this:

<pre>
[cloud-user@icinga2 ~]$ lsb_release -i -s
CentOS
</pre>

Apparently in some cases it does turn into CentOS

File: icinga2x/Vagrantfile (hunk 1 above)
Comment (martbhell) [inline icinga2x/Vagrantfile]: Ah. I could just remove the vm.box for openstack and it'll go through.
I'll amend and add this back for virtualbox.

Comment (martbhell) [inline icinga2x/Vagrantfile]: No actually I get some errors with it. Basically for openstack (I think, might be wrong) I don't want to download a box. Anyway, I can still remove the box option for openstack.

https://github.com/martbhell/icinga-vagrant/blob/05d44cc3e94155ce72a00d1a1dbe91faf1eb0535/icinga2x/Vagrantfile#L89 has an override so that sets it for virtualbox provider. Isn't that enough?

File: icinga2x/Vagrantfile (hunk 2 above)
Comment (martbhell) [inline icinga2x/Vagrantfile]: Hm. I think that was a leftover from some playing around. 
I removed it from both openstack and virtualbox provider and the openstack provider still works ( I can ssh in )

File: modules/icingaweb2/templates/kickstart.ini.erb (hunk 3 above)
Comment (martbhell) [inline modules/icingaweb2/templates/kickstart.ini.erb]: Np, moved into a separate commit this time.

**Comment by martbhell** [comment]
Thanks for the review!

I did an amend/force push to make some more changes.
I don't like the ENV variables, but that's what the openstack client uses.

**Comment by dnsmichi** [comment]
I did not forget about this PR, it just collides with all the other changes going on. 

Some notes

- the Director module is now managed via puppet-icingaweb2, there's no extra node_name magic needed or config files
- Verification against the :net addition in the Vagrantfile is needed, and if Vagrantfile.nodes can be used here for supplying the environment vars.

**Comment by martbhell** [comment]
Same here.

I tried again and basically I have to make the same changes to Vagrant file as a year ago. I did not bother with a rebase, details: - https://github.com/Icinga/icinga-vagrant/compare/master...martbhell:openstack_provider2

It installs and configures things. Things aren't working correctly, but they aren't for my Virtualbox test either.

With the fork/branch above:
 * Both with Virtualbox and openstack provisioned standalone aren't perfect but puppet completes successfully (can't view host details or deploy director configs) - haven't looked further. Maybe they are because I'm running virtualbox in Windows 10 and vagrant in a Windows WSL Ubuntu.
 * Did I test with :net ?
 * Also Vagrantfile.nodes perhaps it can be used, but there's no example in the Vagrantfile.nodes how to add these extra on a per provider basis. How do you want it done?

**Comment by dnsmichi** [comment]
Thanks :) I'll figure something out after Icinga Camp Berlin.

**Comment by martbhell** [comment]
it's almost always after(or before) an Icinga camp! Hoping to go next year :)

**Comment by dnsmichi** [comment]
Picked the provisioner patch with CentOS into another patch (testing Puppet 5). Time will tell how far I get with merging yours - my talk isn't yet finished :-)

**Comment by dnsmichi** [comment]
A short update here: I am hoping to get access and time to the NETWAYS OpenStack platform soon.

File: modules/icingaweb2/templates/kickstart.ini.erb (hunk 3 above)
Comment (dnsmichi) [reply modules/icingaweb2/templates/kickstart.ini.erb]: This was replaced with the puppet-icingaweb2 and module handling.

Comment (dnsmichi) [reply modules/icingaweb2/templates/kickstart.ini.erb]: This was replaced with the puppet-icingaweb2 and module handling.

**Comment by dnsmichi** [comment]
I've learned that Vagrant >= 2.0 uses SMB per default on macOS, so you need to force this to rsync.

https://github.com/puphpet/puphpet/pull/2805

**Comment by dnsmichi** [comment]
Hahahaha this is fucking awesome :-D 

screen shot 2018-10-30 at 15 13 48
screen shot 2018-10-30 at 15 25 22
screen shot 2018-10-30 at 15 26 09
screen shot 2018-10-30 at 15 29 01


Now there's one problem though - the Hiera assigned IP address for this VM is wrong, which is why tools like Graphite don't work. I need to overhaul this a bit.

Thanks to you I now learned Openstack in a day :)

**Comment by martbhell** [comment]
Very nice! Hope the original methods of provisioning still works too :)

**Comment by dnsmichi** [comment]
Yeah, that's my challenge now :) I need to figure out how I can override the synced hieradata folder for the OpenStack floating ip assigned to this instance. Then the problem with missing Graphite graphs etc. can be solved.

**Comment by dnsmichi** [comment]
I do have my own local patch from your implementation offline, I'll pack that into a separate PR including some more documentation. I've figured out a lot in Horizon while collecting the credentials.

**Comment by dnsmichi** [comment]
Switching between providers requires you to unset the specific values from the environment, especially OS_SSH_USER conflicts with the Vagrant default then.

**Comment by dnsmichi** [comment]
Ok, this was a problem with using `node_config` instead of the scoped `override` variable.

```
+      # provider: openstack
+      node_config.vm.provider :openstack do |os, override|
+        override.ssh.username  = "#{ENV['OS_SSH_USER']}"
+        override.ssh.private_key_path = "#{ENV['OS_SSH_PRIVATE_KEY_PATH']}"
```

solves this. Newer Vagrant versions also required me to set the private key path accordingly.

**Comment by dnsmichi** [comment]
I also understand the need for environment variables with OpenStack, there's not really a point with using a static file here.

**Comment by dnsmichi** [comment]
I've found a way to reliably fetch the floating IP. Actually that's the thing written into `ssh_ip` printed above already. This thing can be used in a provisioner script which overrides /hieradata for this node :)

screen shot 2018-10-31 at 10 48 28

**Comment by dnsmichi** [comment]
```
/hieradata/node/icinga2.vagrant.demo.icinga.com.yaml

node::ipaddress:        "185.11.255.22"
node::name:             "icinga2.vagrant.demo.icinga.com"
```

**Comment by dnsmichi** [comment]
Ok, OpenStack doesn't use my specified FQDN from Hiera as node, but uses `icinga2.novalocal` for instance. This must be reflected with a generated YAML configuration file then. This is also the reason why this didn't work reliably on first run :)

```
[root@icinga2 ~]# cat /etc/icinga2/features-enabled/graphite.conf
# This file is managed by Puppet. DO NOT EDIT.
library "perfdata"


object GraphiteWriter "graphite"  {
  host = "185.11.255.22"
  port = 2003
  host_name_template = "icinga2.$host.name$.host.$host.check_command$"
  service_name_template = "icinga2.$host.name$.services.$service.name$.$service.check_command$"
  enable_send_thresholds = false
  enable_send_metadata = false
}
```

**Comment by dnsmichi** [comment]
Closing in favour of #170 - still users should read its history, so putting this onto 2.0.0 milestone.
